src/lib/Generalize.hs

```diff
@@ -0,0 +1,231 @@
+type Generalized (e::E) (n::S) = (Abs (Nest Binder) e n, [Atom n])
+
+generalizeIxDict :: EnvReader m => Dict n -> m n (Generalized Dict n)
+generalizeIxDict dict = liftGeneralizerM do
+  dict' <- sinkM dict
+  dictTy <- getType dict'
+  dictTyGeneralized <- generalizeType dictTy
+  liftEnvReaderM $ generalizeDict dictTyGeneralized dict'
+{-# INLINE generalizeIxDict #-}
+
+generalizeArgs ::EnvReader m => Nest PiBinder n l -> [Atom n] -> m n (Generalized (ListE Atom) n)
+generalizeArgs reqTys allArgs =  liftGeneralizerM $ runSubstReaderT idSubst do
+  PairE (Abs reqTys' UnitE) (ListE allArgs') <- sinkM $ PairE (Abs reqTys UnitE) (ListE allArgs)
+  ListE <$> go reqTys' allArgs'
+  where
+    go ::Nest PiBinder i i' -> [Atom n] -> SubstReaderT AtomSubstVal GeneralizerM i n [Atom n]
+    go Empty [] = return []
+    go (Nest (PiBinder b ty _) bs) (arg:args) = do
+      ty' <- substM ty
+      arg' <- case ty' of
+        TyKind   -> liftSubstReaderT $ generalizeType arg
+        DictTy _ -> generalizeDict ty' arg
+        _ -> return arg
+      args'' <- extendSubst (b@>SubstVal arg') $ go bs args
+      return $ arg' : args''
+    go _ _ = error "zip error"
+{-# INLINE generalizeArgs #-}
+
+-- === generalizer monad plumbing ===
+
+data GeneralizationEmission n l = GeneralizationEmission (Binder n l) (Atom n)
+type GeneralizationEmissions = RNest GeneralizationEmission
+
+newtype GeneralizerM n a = GeneralizerM {
+  runGeneralizerM' :: DoubleInplaceT Env GeneralizationEmissions UnitB HardFailM n a }
+  deriving (Functor, Applicative, Monad, MonadFail, Fallible, ScopeReader, EnvReader, EnvExtender)
+
+liftGeneralizerM
+  :: EnvReader m
+  => (forall l. DExt n l => GeneralizerM l (e l))
+  -> m n (Generalized e n)
+liftGeneralizerM cont = do
+  env <- unsafeGetEnv
+  Distinct <- getDistinct
+  Abs emissions (DoubleInplaceTResult UnitB e) <- return $ runHardFail $
+    runDoubleInplaceT env $ runGeneralizerM' cont
+  let (bs, vals) = hoistGeneralizationVals (unRNest emissions)
+  return (Abs bs e, vals)
+  where
+    -- TODO: something not O(N^2)
+    hoistGeneralizationVals :: Nest GeneralizationEmission n l -> (Nest Binder n l, [Atom n])
+    hoistGeneralizationVals Empty = (Empty, [])
+    hoistGeneralizationVals (Nest (GeneralizationEmission b val) bs) = do
+      let (bs', vals) = hoistGeneralizationVals bs
+      case hoist b (ListE vals) of
+        HoistSuccess (ListE vals') -> (Nest b bs', val:vals')
+        HoistFailure _ -> error "should't happen"
+{-# INLINE liftGeneralizerM #-}
+
+-- XXX: the supplied type may be more general than the type of the atom!
+emitGeneralizationParameter :: Type n -> Atom n -> GeneralizerM n (AtomName n)
+emitGeneralizationParameter ty val = GeneralizerM do
+  Abs b v <- return $ newName noHint
+  let emission = Abs (RNest REmpty (GeneralizationEmission (b:>ty) val)) v
+  emitDoubleInplaceTHoisted emission >>= \case
+    Nothing -> error $ "expected atom to be hoistable " ++ pprint val
+    Just v' -> return v'
```

Comment:
This can't fail because the bottom emissions are UnitB, right?

Reply:
No, it can fail. The hoisted emissions don't just hoist about the bottom emissions (which are UnitB, as you said. They also hoist above every local binder. I think that only includes pi binders in this case, since we're just traversing a type.

Reply:
Ok, good point. Can you please add that as a comment?

Reply:
In writing the comment, I just realized that this only works if the binders within the type are only used in DataParam roles! This was always the intention (e.g. we assume that sort of thing in creating dests for dependent pairs in Imp). But we should make it more official. I'll add it to the issue about listing the remaining checks needed.